File: c/11.盛最多水的容器.c
/*
 * @lc app=leetcode.cn id=11 lang=c
 *
 * [11] 盛最多水的容器
 */

// @lc code=start

#define MIN(a,b) ((b) < (a) ? (b) : (a) )
#define MAX(a,b) ((b) > (a) ? (b) : (a) )
int maxArea(int* height, int heightSize) {
    int ans = 0;
    int left = 0;
    int right = heightSize - 1;


    while (left < right)
    {
        int area = MIN(height[left],height[right]) * (right - left);
        ans = MAX(area,ans);

        height[left] < height[right]? left++ : right--;
    }

    return ans;    
}
// @lc code=end

File: python3/11.盛最多水的容器.py
#
# @lc app=leetcode.cn id=11 lang=python3
#
# [11] 盛最多水的容器
#

# @lc code=start
class Solution:
    def maxArea(self, height: List[int]) -> int:

# 1. 向内移动短板
# 2. 使用双指针
        i = 0
        j = len(height) - 1
        res = 0
    
        while i < j :

            res = max(res, min(height[i], height[j]) * (j - i))
            
            i += 1 if height[i] < height[j] else j -= 1
            
        return res
    
# @lc code=end

